packages/react-portal/Spec.md

```diff
@@ -0,0 +1,158 @@
+
+<ThemeProvider>
+    <Portal className={styles.portalContent>
+      Can use all theme CSS variables from the parent ThemeProvider
+    </Portal>
+</ThemeProvider>
+```
+
+## Variants
+
+- Mounting the portal on a custom node
+
+## API
+
+### Portal
+
+| Name                 | Description                                                   | Required | Type              | Default value                    |
+| -------------------- | ------------------------------------------------------------- | -------- | ----------------- | -------------------------------- |
+| mountNode            | Where the portal is mounted to the DOM                        | No       | HTMLElement       | ProviderContext or document.body |
+
+
+## Structure
+
+```
+<FluentProvider
+  <Portal id="portal-1" />
+  <Portal id="portal-2" />
+</FluentProvider
+```
+
+DOM output:
+```tsx
+<body>
+  <div>Maintree</div>
+
+  <div id="portal-1" class="theme-provider-0"}>{children}</div>
+  <div id="portal-2" class="theme-provider-0"}>{children}</div>
+</body>
+```
+
+## Migration
+
+_Describe what will need to be done to upgrade from the existing implementations:_
+
+### v8 migration
+
+- There will be no way to disable event bubbling, it will be up to consumers to call `stopPropagation` themselves or create extra utilities that do so
+- No more concept of `LayerHost` and id/class selectors, raw HTML elements/refs can be stored in context on the consumer app and used in `mountNode` for `Portals` if required
+- No more mount lifecycle methods, users can remedy this easily with `useEffect` or `useLayoutEffect` hooks
+- `insertFirst` will no longer be supported, and can be handled by a custom `mountNode` if necessary, sticky Dialog can be implmented with `pointer-events: none`
+
+### v0 migration
+
+- No more openable portals - should use future converged `Popup`
+- No more focus trapping in `Portals` do that manually (Tabster)
+- No more mount lifecycle methods, users can remedy this easily with `useEffect` or `useLayoutEffect` hooks
+
+## Behaviors
+
+### Server Side Rendering (SSR)
+
+The ReactDOM `createPortal` requires a valid DOM node to render. This is problematic when `document` does not actually exist during the server render. Instead during the server render `null` will be used. This is not a big problem for most components that use portals such as popups or dialogs since they must be opened from some kind of trigger element (i.e. button)
+
+However, there are some cases where a `Portal` content will always need to be rendered on the page. Tooltips should always be rendered so that `aria` attributes will refer to actual elements. This is problematic since the Tooltip (or higher order component) needs to be aware of the server render where `null`is rendered and render the actual content on the first client render.
+
+The `Portal` component should handle this SSR case, and should be aware of the server and client renders when calling `ReactDOM.createPortal`.
```

Comment:
Sorry if I'm missing it, but I don't see details about how Portal will handle SSR? It seems like we need two different behaviors: one that renders the content inline in SSR, and one that doesn't render anything.

One option would be a prop on Portal. But it seems like the more common/desirable case is to render nothing in SSR, so maybe just leave it up to the component to handle inline rendering in SSR. E.g.:
```typescript
{isSSR() ?
  <Tooltip /> /* inline in SSR */ :
  <Portal><Tooltip /></Portal> /* in a portal normally */}
```

Reply:
So the broad requirement is that `Portal` should be smart enough to handle this IMO, that's the scope of the spec. As a consumer you should be safe knowing that if you use a fluent component, it should be SSR safe. 

On that note I've got an RFC in the pipeline with a concrete solution, stay tuned